auto_dev/protocols/adapters.py

```diff
@@ -0,0 +1,176 @@
+"""Module containing adapter classes for proto_schema_parser."""
+
+from __future__ import annotations
+
+import re
+from dataclasses import field, dataclass
+
+from proto_schema_parser.ast import (
+    Enum,
+    File,
+    Field,
+    Group,
+    OneOf,
+    Import,
+    Option,
+    Comment,
+    Message,
+    Package,
+    Service,
+    MapField,
+    Reserved,
+    Extension,
+    FileElement,
+    ExtensionRange,
+    MessageElement,
+)
+
+
+def camel_to_snake(name: str) -> str:
+    """Convert CamelCase to snake_case."""
+    return re.sub(r"(?<!^)(?=[A-Z])", "_", name).lower()
+
+
+@dataclass
+class MessageAdapter:
+    """MessageAdapter for proto_schema_parser ast.Message."""
+
+    file: FileAdapter | None = field(repr=False)
+    parent: FileAdapter | MessageAdapter | None = field(repr=False)
+    wrapped: Message = field(repr=False)
+    fully_qualified_name: str
+    elements: list[MessageElement | MessageAdapter] = field(default_factory=list, repr=False)
+
+    comments: list[Comment] = field(default_factory=list)
+    fields: list[Field] = field(default_factory=list)
+    groups: list[Group] = field(default_factory=list)
+    oneofs: list[OneOf] = field(default_factory=list)
+    options: list[Option] = field(default_factory=list)
+    extension_ranges: list[ExtensionRange] = field(default_factory=list)
+    reserved: list[Reserved] = field(default_factory=list)
+    messages: list[MessageAdapter] = field(default_factory=list)
+    enums: list[Enum] = field(default_factory=list)
+    extensions: list[Extension] = field(default_factory=list)
+    map_fields: list[MapField] = field(default_factory=list)
```

Comment:
This adapter allows for more convenient processing of message elements, without having to iterate them repeatedly to filter out those you need as a given point. Moreover, reference to the parent allows one to traverse back up the tree to find out where a `Field.type` or `MapField.value_type` is referenced in order to obtain a fully qualified type (function for that in formatter.py). The `file` attribute could likely be removed, it is a remnant from before I realized that this is insufficient and that one in fact need to traverse parents instead.